File: bindings/node/ts/backend.ts
/**
 * The contract both engines implement.
 *
 * There are two: the Node-API addon built by `src/lib.rs`, and the emscripten
 * WebAssembly module in `@sonejs/sone-wasm`. They expose the same method names
 * and the same shapes — document in, bytes out — so nothing above this file
 * ever branches on which one is loaded.
 */

/** A live engine: owns a font registry and an asset cache. */
export interface NativeEngine {
  registerFont(name: string, data: Uint8Array): void;
  registerFontFile(name: string, path: string): void;
  unregisterFont(name: string): void;
  registerImage(name: string, data: Uint8Array): void;
  hasFont(name: string): boolean;
  fontFamilies(): string[];
  resetFonts(): void;
  render(
    document: string,
    format: string,
    density?: number | null,
    quality?: number | null,
    strict?: boolean | null,
  ): Promise<Uint8Array>;
  renderPages(
    document: string,
    format: string,
    density?: number | null,
    quality?: number | null,
    strict?: boolean | null,
  ): Promise<Uint8Array[]>;
  dumpLayout(document: string): Promise<string>;
  dumpMetadata(document: string, granularity?: string | null): Promise<string>;
}

export interface Backend {
  Engine: new (baseDir?: string) => NativeEngine;
  version(): string;
  /** Whether the engine can read files itself. False under WebAssembly. */
  hasFilesystem: boolean;
}

File: bindings/jvm/sone-core/src/main/java/dev/sone/Backend.java
package dev.sone;

import java.util.List;

/**
 * What an engine implementation has to provide.
 *
 * <p>Two exist, and which one loads is a platform fact rather than a choice.
 * Desktop JVMs get {@code dev.sone.panama.PanamaBackend}, which calls the C ABI
 * through {@code java.lang.foreign}. Android has no such package — ART does not
 * implement Panama, and it cannot be desugared because it needs VM support — so
 * Android gets {@code dev.sone.jna.JnaBackend}, which reaches the same C ABI
 * through JNA.
 *
 * <p>Everything above this interface is shared: the whole builder, the property
 * interfaces, the JSON writer and {@link Rendering} never learn which one is
 * underneath.
 */
public interface Backend extends AutoCloseable {

    void registerFont(String name, byte[] data);

    void registerFontFile(String name, String path);

    void registerImage(String name, byte[] data);

    boolean hasFont(String name);

    List<String> fontFamilies();

    void resetFonts();

    byte[] render(String document, OutputFormat format, Double density, double quality, boolean strict);

    List<byte[]> renderPages(String document, OutputFormat format, Double density, double quality, boolean strict);

    String dumpLayout(String document);

    String dumpMetadata(String document, Granularity granularity);

    /** The native library version. */
    String version();

    @Override
    void close();
}
